cartography/intel/aws/ec2/route_tables.py

```diff
@@ -0,0 +1,520 @@
+
+
+@timeit
+def load_associations(
+    neo4j_session: neo4j.Session, data: List[Dict], region: str,
+    aws_account_id: str, update_tag: int,
+) -> None:
+    ingest_subnet_associations = """
+    UNWIND {route_tables} AS route_table
+    UNWIND route_table.Associations AS assoc
+
+    MERGE (rt:EC2RouteTable{id: route_table.RouteTableId})
+    ON CREATE SET rt.firstseen = timestamp()
+    SET rt.lastupdated = {update_tag},
+    rt.route_table_id = route_table.RouteTableId
+
+    MERGE (rta:EC2RouteTableAssociation{id: assoc.RouteTableAssociationId})
+    ON CREATE SET rta.firstseen = timestamp()
+    SET rta.state = assoc.AssociationState.state,
+    rta.subnet_id = assoc.SubnetId,
+    rta.main = assoc.Main,
+    rta.lastupdated = {update_tag}
+
+    MERGE (rta)-[rta_a:ASSOCIATED]->(rt)
+    ON CREATE SET rta_a.firstseen = timestamp()
+    SET rta_a.lastupdated = {update_tag}
+
+
+    FOREACH(runMe IN CASE WHEN assoc.SubnetId IS NOT NULL THEN [1] ELSE [] END |
+
+        MERGE (rt:EC2RouteTable{id: route_table.RouteTableId})
+        ON CREATE SET rt.firstseen = timestamp()
+        SET rt.lastupdated = {update_tag},
+        rt.route_table_id = route_table.RouteTableId
+
+        MERGE (subnet:EC2Subnet{subnetid: assoc.SubnetId})
+        ON CREATE SET subnet.firstseen = timestamp()
+        SET subnet.lastupdated = {update_tag}
+
+        MERGE (subnet)-[r:ASSOCIATED]->(rt)
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = {update_tag}, r.main = assoc.Main,
+        r.state = assoc.AssociationState.State,
+        r.id = assoc.RouteTableAssociationId
+
+        MERGE (subnet)-[s_rta:ASSOCIATED]->(rta)
+        ON CREATE SET s_rta.firstseen = timestamp()
+        SET s_rta.lastupdated = {update_tag}
+
+    )
+
+
+    """
+
+    neo4j_session.run(
+        ingest_subnet_associations, route_tables=data, update_tag=update_tag,
+        region=region, aws_account_id=aws_account_id,
+    )
+
+
+@timeit
+def load_default_associations(
+    neo4j_session: neo4j.Session, data: List[Dict], region: str,
+    aws_account_id: str, update_tag: int,
+) -> None:
+    ingest_default_associations = """
+
+    MATCH (a:AWSAccount{id:{aws_account_id}})-[:RESOURCE]-(snet:EC2Subnet)-[]-(vpc:AWSVpc)-
+    [:RESOURCE]-(rtb:EC2RouteTable)-[:ASSOCIATED]-(rtba:EC2RouteTableAssociation{main:true})
+    WHERE NOT (snet)-[:ASSOCIATED]-(:EC2RouteTable)
+
+    MERGE (snet)-[r_m:ASSOCIATED]->(rtb)
+    ON CREATE SET r_m.firstseen = timestamp()
+    SET r_m.lastupdated = {update_tag}, r_m.main = rtba.Main,
+    r_m.state = rtba.AssociationState.State,
+    r_m.id = rtba.RouteTableAssociationId
+
+    MERGE (subnet)-[s_rta_m:ASSOCIATED]->(rtba)
+    ON CREATE SET s_rta_m.firstseen = timestamp()
+    SET s_rta_m.lastupdated = {update_tag}
+    """
+
+    neo4j_session.run(
+        ingest_default_associations, route_tables=data, update_tag=update_tag,
+        region=region, aws_account_id=aws_account_id,
+    )
+
+
+@timeit
+def load_routes(
```

Comment:
Ideally, rather than have neo4j itself perform control flow, we could try to transform the data into simpler lists of dicts that can be merged in a single `unwind`. Is there a way to ensure all items in a list are the same type of gateway to make things simpler?

But yeah in general, let's try to have as few merges per query as possible because we've found that neo4j can throw unpredictable errors when we throw enough data at it and if queries aren't exactly the way it wants them; I write about this a bit in https://github.com/lyft/cartography/pull/631 (in progress; let me know what you think there too).

Reply:
:thinking: I thought we optimize on not processing the data too much before loading to neo4j. This would mean to rewrite the most complex part of the PR.

Reply:
@krisek - My main reason for being hesitant on the long query at first was because Neo4j would throw ServiceUnavailable exceptions that were difficult to repro and understand. Now that I have a [viable fix](https://github.com/lyft/cartography/issues/522#issuecomment-885402304) for it when it happens, I'm ok with the query form as long as we call `neo4j_session.write_transaction()` so that we can get automatic retries in case the database connection times out for any reason.

Here's an example:

```python
def _load_findings_tx(tx: neo4j.Transaction, findings: List[Dict], update_tag: int) -> None:
    query = """
        query goes here
    """
    logger.info(f"Loading {len(findings)} findings to the graph")
    tx.run(query, Findings=findings, UpdateTag=update_tag)


def load(session: neo4j.Session, findings: LIst[Dict]) -> None:
    session.write_transaction(_load_findings_tx, findings, update_tag)
```

As long as load functions call queries this way, I think we'll avoid the exception more effectively - I'll update the intel module guide at some point.